### ISSUE TITLE
Question about implementation

### ISSUE THREAD

**Issue by tkuanlun350** [ppwwyyxx/tensorpack#594]
Hi ! 
I am currently trying to add FPN network in your FasterRCNN codebase but encountered some problems. I will be very grateful if you can share some thoughts !

(1. I am facing CUDNN_STATUS_BAD_PARAM seemed to be caused by empty array in cudnn. But I have already use tf 1.4.0 and it should be fixed by your pull request right ?

(2. I have successfully trained FPN with RPN (loss and prediction seemed to be reasonable) and now
I am implementing Pyramid ROI align by iterate all strides and determine the level of roi  to pool according to its size. My question is that I have to preserve original order of boxes in order to match the sampled rois (in function sample_fast_rcnn_targets). I am trying to record box indices and use scatter_nd to recover original order. But it will cause slightly shape difference in decode_bbox_target in runtime (and also problem #1 when I simply discard tf.scatter_nd to return pooled feature).

The fraction of the code is like below:
```
def roi_align_FPN(featuremaps, boxes, output_shape):    
    # feature map [P5, P4, P3, P2]
    boxes = tf.stop_gradient(tf.expand_dims(boxes, 0)) 
    x1, y1, x2, y2 = tf.split(boxes, 4, axis=2) 
    w = x2 - x1
    h = y2 - y1
    roi_level = tf_log2(tf.sqrt(h * w) / (224.0))
    roi_level = tf.minimum(5, tf.maximum(
            2, 4 + tf.cast(tf.round(roi_level), tf.int32)))
    roi_level = tf.squeeze(roi_level, 2)
    # limit to P5 ~ P2
    print("ROI LEVEL SHAPE:", roi_level.get_shape())
    print("BOX SHAPE", boxes.get_shape())
    pooled = []
    box_to_level = []
    strides = [4., 8., 16., 32.]
    for level in range(2, 6):
        featuremap_to_crop = featuremaps[4 - (level - 2)]
        # order : P6(idx 4) ~ P2(idx0) => idx shift
        # P2 -> f[4], P3 -> f[3], P4 -> f[2], P5 -> f[1]
        id_for_box_wrt_level = tf.where(tf.equal(roi_level, level))
        level_boxes = tf.gather_nd(boxes, id_for_box_wrt_level)
        # Box indicies for crop_and_resize.
        box_indices = tf.cast(id_for_box_wrt_level[:, 0], tf.int32)
        box_to_level.append(tf.cast(id_for_box_wrt_level[:, 1], tf.int32))

        level_boxes = tf.stop_gradient(level_boxes)
        box_indices = tf.stop_gradient(box_indices)
        # Perform roi pool (current box scale still wrt image size)
        level_boxes = level_boxes * (1.0 / strides[level - 2])
        ret = crop_and_resize(
                featuremap_to_crop, level_boxes,
                tf.zeros([tf.shape(level_boxes)[0]], dtype=tf.int32),
                output_shape * 2)
        ret = tf.nn.avg_pool(ret, [1, 1, 2, 2], [1, 1, 2, 2], padding='SAME', data_format='NCHW')
        pooled.append(ret)

    pooled = tf.concat(pooled, axis=0)
    box_to_level = tf.concat(box_to_level, axis=0)
    pooled = tf.scatter_nd(box_to_level, pooled, tf.shape(pooled), name="scatter_pool")
```

Thanks !

**Comment by ppwwyyxx** [comment]
1. It's in 1.5
2. I don't understand what is the "shape difference" you're talking about. I don't have the context about what's going on in your code. But this sounds like a general tensorflow question to me. Are you talking about a potential bug in your code that you couldn't find or what?

**Comment by tkuanlun350** [comment]
Turns out it is not related to the cudnn bugs. Thanks for your help. The rest of the problem should better ask on stackoverflow.

**Comment by qinhuan** [comment]
@tkuanlun350  Hi, Have you finished the code of FPN? Could you share the code? Thanks.

**Comment by tkuanlun350** [comment]
@qinhuan Yes, but performance dropped to 0.39 using 4 gpus. There maybe some bugs in my implementation :(

**Comment by ppwwyyxx** [comment]
FYI, FPN added in examples: https://github.com/ppwwyyxx/tensorpack/tree/master/examples/FasterRCNN

**Comment by tkuanlun350** [comment]
@ppwwyyxx  The solution to tf.image.imresize is awesome. Thanks! great job.
I still have an question about fpn's  rpn loss, if you can kindly explain to me will be very appreciated.
May I ask you why rpn loss is computed level-wise and what is the meaning of placeholder in rpn loss?

**Comment by ppwwyyxx** [comment]
> why rpn loss is computed level-wise

I'm not sure I understand what you're asking. What is the alternative if not level-wise? Each level has different image sizes, so I compute loss on each level separately.

When the loss does not exist (no boxes at all) I don't want it to be shown as NaN. A placeholder can be set to any number and it will not affect training. It just makes the logs and tensorboard look nicer.

**Comment by ppwwyyxx** [comment]
If the loss is accumulated over all the valid anchors, that sounds equivalent to what is in my code.
